distributed/tests/test_spans.py

```diff
@@ -258,35 +327,103 @@ def update_graph(self, scheduler, *, tasks, annotations, **kwargs):
     # First task to attach to the TaskGroup sets the span. This is arbitrary.
-    assert spans.keys() == {span_id}
-    assert len(spans[span_id].groups) == 1
-    assert s.task_groups["x"].span == span_id
+    assert sbn.keys() == {span_name}
+    assert len(sbn[span_name][0].groups) == 1
+    assert s.task_groups["x"].span_id == sbn[span_name][0].id
 
     sts0 = s.tasks[str(x0.key)]
     sts1 = s.tasks[str(x1.key)]
     wts0 = a.state.tasks[str(x0.key)]
     wts1 = a.state.tasks[str(x1.key)]
-    assert sts0.group == sts1.group
-    # The annotation on x1 has been changed
-    if use_default:
-        assert not sts0.annotations
-        assert not sts1.annotations
-        assert not wts0.annotations
-        assert not wts1.annotations
-        assert s.plugins["my_plugin"].seen == [(["('x', 0)"], {}), (["('x', 1)"], {})]
-    else:
-        assert (
-            sts0.annotations
-            == sts1.annotations
-            == wts0.annotations
-            == wts1.annotations
-            == ({"span": ("p1",)})
-        )
-        assert s.plugins["my_plugin"].seen == [
-            (["('x', 0)"], {"span": {"('x', 0)": ("p1",)}}),
-            (["('x', 1)"], {"span": {"('x', 1)": ("p1",)}}),
-        ]
+    assert sts0.group is sts1.group
+    assert wts0.span_id == wts1.span_id
+
+
+def test_no_tags():
+    with pytest.raises(ValueError, match="at least one"):
+        with span():
+            pass
+
+
+@gen_cluster(client=True)
+async def test_client_desires_new_key_creates_ts(c, s, a, b):
+    """A TaskState object is created by client_desires_keys, and
+    is only later submitted with submit/compute by a different client
+
+    See also
+    --------
+    test_scheduler.py::test_client_desires_new_key_creates_ts
+    test_spans.py::test_client_desires_new_key_creates_tg
+    test_spans.py::test_scatter_creates_ts
+    test_spans.py::test_scatter_creates_tg
+    """
+    x = Future(key="x")
+    await wait_for_state("x", "released", s)
+    assert s.tasks["x"].group.span_id is None
+    async with Client(s.address, asynchronous=True) as c2:
+        c2.submit(inc, 1, key="x")
+        assert await x == 2
+    assert s.tasks["x"].group.span_id is not None
+
+
+@gen_cluster(client=True)
+async def test_client_desires_new_key_creates_tg(c, s, a, b):
+    """A TaskGroup object is created by client_desires_keys, and
+    only later gains runnable tasks
+
+    See also
+    --------
+    test_spans.py::test_client_desires_new_key_creates_ts
+    test_spans.py::test_scatter_creates_ts
+    test_spans.py::test_scatter_creates_tg
+    """
+    x0 = Future(key="x-0")
+    await wait_for_state("x-0", "released", s)
+    assert s.tasks["x-0"].group.span_id is None
+    x1 = c.submit(inc, 1, key="x-1")
+    assert await x1 == 2
+    assert s.tasks["x-0"].group.span_id is not None
+
+
+@gen_cluster(client=True)
+async def test_scatter_creates_ts(c, s, a, b):
+    """A TaskState object is created by scatter, and only later becomes runnable
```

Comment:
OOT: It's mildly scary that this is even legitimate behavior.

Reply:
To be frank I think that the whole scatter design and implementation should be levelled and redone from scratch